source/common/upstream/load_balancer_impl.cc

```diff
@@ -36,24 +28,81 @@ static const std::string RuntimeZoneEnabled = "upstream.zone_routing.enabled";
 static const std::string RuntimeMinClusterSize = "upstream.zone_routing.min_cluster_size";
 static const std::string RuntimePanicThreshold = "upstream.healthy_panic_threshold";
 
+uint32_t LoadBalancerBase::choosePriority(uint64_t hash,
+                                          const std::vector<uint32_t>& per_priority_load) {
+  hash = hash % 100 + 1; // 1-100
+  uint32_t aggregate_percentage_load = 0;
+  // As with tryChooseLocalLocalityHosts, this can be refactored for efficiency
+  // but O(N) is good enough for now given the expected number of priorities is
+  // small.
+  for (size_t priority = 0; priority < per_priority_load.size(); ++priority) {
+    aggregate_percentage_load += per_priority_load[priority];
+    if (hash <= aggregate_percentage_load) {
+      return priority;
+    }
+  }
+  // The percentages should always add up to 100 but we have to have a return for the compiler.
+  ASSERT(0);
+  return 0;
+}
+
 LoadBalancerBase::LoadBalancerBase(const PrioritySet& priority_set, ClusterStats& stats,
                                    Runtime::Loader& runtime, Runtime::RandomGenerator& random)
-    : stats_(stats), runtime_(runtime), random_(random), priority_set_(priority_set),
-      best_available_host_set_(bestAvailable(&priority_set)) {
-  per_priority_load_.resize(priority_set.hostSetsPerPriority().size());
-  per_priority_load_[best_available_host_set_->priority()] = 100;
-  priority_set_.addMemberUpdateCb([this](uint32_t, const std::vector<HostSharedPtr>&,
-                                         const std::vector<HostSharedPtr>&) -> void {
-    per_priority_load_.resize(priority_set_.hostSetsPerPriority().size());
-    per_priority_load_[best_available_host_set_->priority()] = 0;
-    // Update the host set to use for picking, based on the new state.
-    best_available_host_set_ = bestAvailable(&priority_set_);
-    // With current picking logic, the best available host set gets 100% of
-    // traffic and all others get 0%
-    per_priority_load_[best_available_host_set_->priority()] = 100;
-  });
+    : stats_(stats), runtime_(runtime), random_(random), priority_set_(priority_set) {
+  for (auto& host_set : priority_set_.hostSetsPerPriority()) {
+    recalculatePerPriorityState(host_set->priority());
+  }
+  priority_set_.addMemberUpdateCb(
+      [this](uint32_t priority, const std::vector<HostSharedPtr>&,
+             const std::vector<HostSharedPtr>&) -> void { recalculatePerPriorityState(priority); });
+}
 
-} // namespace Upstream
+void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority) {
+  per_priority_load_.resize(priority_set_.hostSetsPerPriority().size());
+  per_priority_health_.resize(priority_set_.hostSetsPerPriority().size());
+
+  // Determine the health of the newly modified priority level.
+  // Health ranges from 0-100, and is the ratio of healthy hosts to total hosts, modified by the
+  // overprovision factor of 1.4
```

Comment:
Can you add some more detail on where 1.4 comes from? Also, nit, period end of sentence.

Reply:
From a decade working somewhere where we over-provision roughly 20%.  I can make it configurable from the get go but I figured we could leave it until someone actually cared.  

I'll comment both that it's arbitrary and can eventually be configurable unless you'd prefer I make it a config option now.